packages/bsky/src/views/index.ts

```diff
@@ -1111,6 +1131,462 @@ export class Views {
+      const post = state.posts?.get(uri)
+      const parentUri = post?.record.reply?.parent.uri
+      if (!parentUri) return
+      if (includedPosts.has(uri)) return
+      includedPosts.add(uri)
+      childrenByParentUri[parentUri] ??= []
+      childrenByParentUri[parentUri].push(uri)
+    })
+
+    const rootUri = getRootUri(anchor, post)
+    const opDid = uriToDid(rootUri)
+    const authorDid = postView.author.did
+    const isOpPost = authorDid === opDid
+    const anchorViolatesThreadGate = post.violatesThreadGate
+
+    // Builds the parent tree, and whether it is a contiguous OP thread.
+    const parentTree = !anchorViolatesThreadGate
+      ? this.threadV2Parent({
+          childUri: anchor,
+          viewer,
+          opDid,
+          rootUri,
+          state,
+          above: opts.above,
+          depth: -1,
+        })
+      : undefined
+
+    const { tree: parent, isOPThread: isOpThreadFromRootToParent } =
+      parentTree ?? { tree: undefined, isOPThread: false }
+
+    const isOPThread = parent
+      ? isOpThreadFromRootToParent && isOpPost
+      : isOpPost
+
+    let anchorTree: ThreadTree
+    if (this.noUnauthenticatedPost(viewer, postView)) {
+      anchorTree = {
+        item: this.threadV2ItemNoUnauthenticated({
+          uri: anchor,
+          depth: 0,
+        }),
+        parent,
+      }
+    } else {
+      anchorTree = {
+        item: this.threadV2ItemPost(
+          {
+            depth: 0, // The depth of the anchor post is always 0.
+            isOPThread,
+            postView,
+            rootUri,
+            uri: anchor,
+            viewer,
+          },
+          state,
+        ),
+        hasOPLike: !!state.threadContexts?.get(postView.uri)?.like,
+        parent,
+        replies: !anchorViolatesThreadGate
+          ? this.threadV2Replies({
+              parentUri: anchor,
+              viewer,
+              isOPThread,
+              opDid,
+              rootUri,
+              childrenByParentUri,
+              state,
+              below: opts.below,
+              depth: 1,
+            })
+          : undefined,
+      }
+    }
+
+    return sortTrimFlattenThreadTree(anchorTree, {
+      opDid,
+      branchingFactor: opts.branchingFactor,
+      sorting: opts.sorting,
+      prioritizeFollowedUsers: opts.prioritizeFollowedUsers,
+      viewer,
+      fetchedAt: Date.now(),
+    })
+  }
+
+  private threadV2Parent({
+    childUri,
+    viewer,
+    opDid,
+    rootUri,
+    state,
+    above,
+    depth,
+  }: {
+    childUri: string
+    viewer: HydrateCtx['viewer']
+    opDid: string
+    rootUri: string
+    state: HydrationState
+    above: number
+    depth: number
+  }): { tree: ThreadTree; isOPThread: boolean } | undefined {
+    // Reached the `above` limit.
+    if (Math.abs(depth) > above) {
+      return undefined
+    }
+
+    // Not found.
+    const uri = state.posts?.get(childUri)?.record.reply?.parent.uri
+    if (!uri) {
+      return undefined
+    }
```

Comment:
One thing we couldn't do in `getPostThread` is include a deleted post but maintain the ancestor chain, e.g.:
```
0         # anchor's parent's parent appears in full above deleted post
0_0✗      # anchor's parent is deleted
0_0_0     # anchor
0_0_0_0   # reply to anchor
```

Might be nice to introduce!  But we can always do it in a future pass.